File: frontend/src/app/variableComponentInfo.ts
import { IdentifiedInfo } from './identifiedInfo';

export class VariableComponentInfo {
    uri: string;
    cellID: string;
    operator: string;
    variants: IdentifiedInfo[] = [];

    constructor(cellID?: string){
        if(cellID)
            this.cellID = cellID;
    }

    makeCopy(): VariableComponentInfo {
        const copy: VariableComponentInfo = new VariableComponentInfo(this.cellID);
        copy.uri = this.uri;
        copy.operator = this.operator;
        for(let variant of this.variants){
            copy.variants.push(variant.makeCopy());
        }
        return copy;
    }

    addVariant(info: IdentifiedInfo){
        // avoid duplicates
        for(let variant of this.variants){
            if(variant.uri == info.uri)
                return;
        }
        this.variants.push(info);
    }
}

File: backend/src/data/VariableComponentInfo.java
package data;

public class VariableComponentInfo {

	private String uri;
	private String cellID;
	private String operator;
	private IdentifiedInfo[] variants;

	public String getUri() {
		return uri;
	}

	public void setUri(String uri) {
		this.uri = uri;
	}

	public String getCellID() {
		return cellID;
	}

	public void setCellID(String cellID) {
		this.cellID = cellID;
	}

	public String getOperator() {
		return operator;
	}

	public void setOperator(String operator) {
		this.operator = operator;
	}

	public IdentifiedInfo[] getVariants() {
		return variants;
	}

	public void setVariants(IdentifiedInfo[] variants) {
		this.variants = variants;
	}

}
